// This file defines types used for reach queries and responses.

syntax = "proto3";

import "google/api/annotations.proto";
import "candig/clinical_metadata.proto";
import "candig/clinical_metadata_service.proto";
import "candig/variants.proto";
import "candig/variant_service.proto";

package candig;


service SearchService {
    rpc GetItem (SearchQueryRequest) 
        returns (SearchQueryResponse) {
            option (google.api.http) = {
                post: "/v0.8.0/search"
                body: "*"
            };
    };
};


// Contains statements exclusively on a single table
message Component {

    // A unique identifier of the current component
    string id = 1;

    oneof endpoint {
    
        SearchPatientsRequest patients = 2;
        
        SearchEnrollmentsRequest enrollments = 3;
        
        SearchConsentsRequest consents = 4;
    
        SearchDiagnosesRequest diagnoses = 5;
        
        SearchSamplesRequest samples = 6;
    
        SearchTreatmentsRequest treatments = 7;
        
        SearchOutcomesRequest outcomes = 8;
        
        SearchComplicationsRequest complications = 9;
        
        SearchTumourboardsRequest tumourboards = 10;
        
        SearchVariantsRequest variants = 11;

        SearchVariantsByGeneNameRequest variantsByGene = 12;

        SearchSlidesRequest slides = 13;

        SearchStudiesRequest studies = 14;

        SearchLabtestsRequest labtests = 15;

        SearchChemotherapiesRequest chemotherapies = 16;

        SearchRadiotherapiesRequest radiotherapies = 17;

        SearchImmunotherapiesRequest immunotherapies = 18;

        SearchSurgeriesRequest surgeries = 19;

        SearchCelltransplantsRequest celltransplants = 20;
    }
}


message Logic {

    repeated Logic and = 1;
    
    repeated Logic or = 2;

    string id = 3;

    bool negate = 4;
}



// Definition of which table and which fields should be returned
message Result {

    // Name of a single table from which fields are requested
    string table = 1;

    // List of fields of the requested table
    repeated string fields = 2;

    // Start bp. Used if the user is requesting variants endpoint
    string start = 3;

    // End bp. Used if the user is requesting variants endpoint
    string end = 4;

    // Reference genome. Used if the user is requesting variants endpoint.
    string referenceName = 5;

    // Gene name. Used if the user is requesting variantsByGene endpoint.
    string gene = 6;

    // List of fields to aggregate by and return counts. Overrides response format if included.
    repeated string count = 7;
}


//Request object for the search function
message SearchQueryRequest {

    // The datasetId to query from
    string dataset_id = 1;

    // Definition of the high order logical operations on components
    Logic logic = 2;

    // Array of single table queries
    repeated Component components = 3;

    // Definition of the requested tables and fields
    repeated Result results = 4;

    // Limit of the maximum number of records
    int32 limit = 5;
    
    // Specifies the maximum number of results to return in a single page.
    // If unspecified, a system default will be used.
    int32 page_size = 6;

    // The continuation token, which is used to page through large result sets.
    // To get the next page of results, set this parameter to the value of
    // `nextPageToken` from the previous response.
    string page_token = 7;    
}


//Response object for the search function
message SearchQueryResponse {

    // Objects from clinical metadata
    
    // List of patients
    repeated Patient patients = 1;
    // List of enrollments
    repeated Enrollment enrollments = 2;
    // List of consents
    repeated Consent consents = 3;
    // List of diagnoses
    repeated Diagnosis diagnoses = 4;
    // List of samples
    repeated Sample samples = 5;
    // List of treatments
    repeated Treatment treatments = 6;
    // List of outcomes
    repeated Outcome outcomes = 7;
    // List of complications
    repeated Complication complications = 8;
    // List of tumourboard results
    repeated Tumourboard tumourboards = 9;

    //Objects from variants

    // List of variants
    repeated Variant variants = 10;

    repeated Slide slides = 11;
    repeated Study studies = 12;
    repeated Labtest labtests = 13;
    repeated Surgery surgeries = 14;
    repeated Chemotherapy chemotherapies = 15;
    repeated Immunotherapy immunotherapies = 16;
    repeated Radiotherapy radiotherapies = 17;
    repeated Celltransplant celltransplants = 18;
    
    // The continuation token, which is used to page through large result sets.
    // Provide this value in a subsequent request to return the next page of
    // results. This field will be empty if there aren't any additional results.
    string next_page_token = 19;
    }
